File: src/main/java/at/meks/PowerDataRepo.java
package at.meks;

import jakarta.enterprise.context.ApplicationScoped;

import java.util.Collection;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@ApplicationScoped
public class PowerDataRepo {

    private final Set<PowerData> dataList = new HashSet<>();

    public void add(PowerData powerData) {
        dataList.add(powerData);
    }

    public List<PowerData> asList(int year) {
        return dataList.stream()
                       .filter(powerData -> powerData.timestampUntil().minusSeconds(1).getYear() == year)
                       .sorted(Comparator.comparing(PowerData::timestampUntil))
                       .toList();
    }

    public Collection<Integer> years() {
        return dataList.stream()
                       .map(powerData -> powerData.timestampUntil().minusSeconds(1).getYear())
                       .distinct()
                       .toList();
    }
}

File: composeApp/src/commonMain/kotlin/at/meks/pv/forecast/battery/powerdata/PowerDataRepo.kt
package at.meks.pv.forecast.battery.powerdata

import at.meks.pv.forecast.battery.Year
import at.meks.pv.forecast.battery.minusSeconds

class PowerDataRepo {

    val dataList = mutableSetOf<SinglePowerData>()

    fun add(singlePowerData: SinglePowerData) {
        dataList.add(singlePowerData)
    }

    fun powerData(year: Year): PowerData {
        return PowerData(dataList
            .filter { powerData -> powerData.timestampUntil.minusSeconds(1).year == year.getValue() }
            .sortedBy(SinglePowerData::timestampUntil)
            .toList())
    }

    fun years(): List<Int> {
        return dataList
            .map { powerData -> powerData.timestampUntil.minusSeconds(1).year }
            .distinct()
            .toList()
    }
}
